src/Microsoft.Extensions.FileSystemGlobbing/InMemoryDirectoryInfo.cs

```diff
@@ -0,0 +1,164 @@
+            if (string.IsNullOrEmpty(rootDir))
+            {
+                throw new ArgumentNullException(nameof(rootDir));
+            }
+
+            if (files == null)
+            {
+                files = new List<string>();
+            }
+
+            Name = Path.GetFileName(rootDir);
+            if (normalized)
+            {
+                _files = files;
+                FullName = rootDir;
+            }
+            else
+            {
+                var fileList = new List<string>(files.Count());
+
+                // normalize
+                foreach (var file in files)
+                {
+                    fileList.Add(Path.GetFullPath(file.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar)));
+                }
+
+                _files = fileList;
+
+                FullName = Path.GetFullPath(rootDir.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar));
+            }
+        }
+
+        /// <inheritdoc />
+        public override string FullName { get; }
+
+        /// <inheritdoc />
+        public override string Name { get; }
+
+        /// <inheritdoc />
+        public override DirectoryInfoBase ParentDirectory
+        {
+            get
+            {
+                return new InMemoryDirectoryInfo(Path.GetDirectoryName(FullName), _files, true);
+            }
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<FileSystemInfoBase> EnumerateFileSystemInfos()
+        {
+            var dict = new Dictionary<string, List<string>>();
+            foreach (var file in _files)
+            {
+                if (!IsRootDirectory(FullName, file))
+                {
+                    continue;
+                }
+
+                var endPath = file.Length;
+                var beginSegment = FullName.Length + 1;
+                var endSegment = file.IndexOfAny(DirectorySeparators, beginSegment, endPath - beginSegment);
+
+                if (endSegment == -1)
+                {
+                    yield return new InMemoryFileInfo(file, this);
+                }
+                else
+                {
+                    var name = file.Substring(0, endSegment);
+                    List<string> list;
+                    if (!dict.TryGetValue(name, out list))
+                    {
+                        dict[name] = new List<string> { file };
+                    }
+                    else
+                    {
+                        list.Add(file);
+                    }
+                }
+            }
+
+            foreach (var item in dict)
+            {
+                yield return new InMemoryDirectoryInfo(item.Key, item.Value, true);
+            }
+        }
+
+        private bool IsRootDirectory(string rootDir, string filePath)
+        {
+            if (!filePath.StartsWith(rootDir, StringComparison.Ordinal)
+                || filePath.IndexOf(Path.DirectorySeparatorChar, rootDir.Length) != rootDir.Length)
```

Comment:
Any reason why? IMO I consider it on the new line to be much more readable and easier to understand. I can immediately see that these two conditions are combined with a logical OR, instead of having to go to the earlier line and look all the way to the right to find the end (and the length various depending on the statement).